api-compose-object.go

```diff
@@ -169,21 +121,21 @@ type SourceInfo struct {
 // `decryptSSEC` is the decryption key using server-side-encryption
 // with customer provided key. It may be nil if the source is not
 // encrypted.
-func NewSourceInfo(bucket, object string, decryptSSEC *SSEInfo) SourceInfo {
+func NewSourceInfo(bucket, object string, sse encrypt.ServerSide) SourceInfo {
 	r := SourceInfo{
 		bucket:     bucket,
 		object:     object,
 		start:      -1, // range is unspecified by default
-		decryptKey: decryptSSEC,
+		encryption: sse,
 		Headers:    make(http.Header),
 	}
 
 	// Set the source header
 	r.Headers.Set("x-amz-copy-source", s3utils.EncodePath(bucket+"/"+object))
 
 	// Assemble decryption headers for upload-part-copy request
-	for k, v := range decryptSSEC.getSSEHeaders(true) {
-		r.Headers.Set(k, v)
+	if r.encryption != nil {
+		encrypt.ServerSideCopy(r.encryption).Marshal(r.Headers)
```

Comment:
I think it is little easier if we write MarshalCopy() function instead of doing like this

Reply:
I thought about this first too - but there are multiple issue (IMO) with the approach:
 1. `*Copy` sounds like we are copying something. But we don't do any copy - bit misleading...
 2. SSE-C is the only form of server-side-encryption which distinguish copy / non-copy.
     So we would pollute the interface only for that...
 3. This is (at the moment) the only place we use the copy construction...